device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceClientManager.java

```diff
@@ -0,0 +1,157 @@
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.DeviceClient;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.experimental.Delegate;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+
+@Slf4j
+public class DeviceClientManager implements IotHubConnectionStatusChangeCallback {
+    private enum ConnectionStatus {
+        DISCONNECTED, CONNECTING, CONNECTED
+    }
+
+    private static final Object lock = new Object();
+    private static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 10;
+    private final boolean autoReconnectOnDisconnected;
+    private ConnectionStatus connectionStatus;
+    @Delegate
+    private DeviceClient client;
+
+    DeviceClientManager(DeviceClient deviceClient, boolean autoReconnectOnDisconnected) {
+        this.connectionStatus = ConnectionStatus.DISCONNECTED;
+        this.client = deviceClient;
+        this.client.registerConnectionStatusChangeCallback(this, this);
+        this.autoReconnectOnDisconnected = autoReconnectOnDisconnected;
+    }
+
+    public void connect() {
+        synchronized (lock) {
+            if(connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    public void disconnect() {
+        synchronized (lock) {
+            try {
+                log.debug("[disconnect] - Closing the device client instance...");
+                client.closeNow();
+            }
+            catch (IOException e) {
+                log.error("[disconnect] - Exception thrown while closing DeviceClient instance: ", e);
+            } finally {
+                connectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+
+    }
+
+    @Override
```

Comment:
override the connection status change listener so that it cannot be set from application